toolchain/check/handle_require.cpp

```diff
@@ -189,21 +198,49 @@ auto HandleParseNode(Context& context, Parse::RequireDeclId node_id) -> bool {
                       "`Self` must appear in the self-type or as a generic "
                       "parameter for each `interface` or `constraint`");
     context.emitter().Emit(node_id, RequireImplsMissingSelf);
+    DiscardGenericDecl(context);
     return true;
   }
 
   if (scope_inst_id == SemIR::ErrorInst::InstId) {
     // `require` is in the wrong scope.
+    DiscardGenericDecl(context);
+    return true;
+  }
+  if (self_inst_id == SemIR::ErrorInst::InstId ||
+      constraint_inst_id == SemIR::ErrorInst::InstId) {
+    // Can't build a useful `require` with an error, it couldn't do anything.
+    DiscardGenericDecl(context);
     return true;
   }
 
   if (identified.required_interfaces().empty()) {
     // A `require T impls type` adds no actual constraints.
+    DiscardGenericDecl(context);
     return true;
   }
 
-  // TODO: Add the `require` constraint to the InterfaceDecl or ConstraintDecl
-  // from `scope_inst_id`.
+  auto require_decl =
+      SemIR::RequireDecl{.type_id = SemIR::TypeType::TypeId,
```

Comment:
What does the type being `type` here mean? Is a `RequireDecl` a type in some sense?